### PR TITLE
Use `global_step` while restoring logging step for old checkpoints

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Here #13467, we fixed the issue by restoring the logging step from the checkpoint, but for the cases where old checkpoints (<v1.6.5) are used, the behavior will still be the same because that key will not be present in old checkpoints and logging will start from 0. To avoid this, the default behavior is changed to restore global_step instead.

Discussed with @carmocca 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

### CODE REVIEW

File: tests/tests_pytorch/models/test_restore.py [hunk 1]
```diff
@@ -255,13 +255,35 @@ class TestModel(BoringModel):
...
+            assert self.trainer.fit_loop.epoch_loop._batches_that_stepped == first_max_epochs * train_batches
 
     trainer.fit(TestModel(), ckpt_path=ckpt_path)
     assert trainer.current_epoch == max_epochs
     assert trainer.global_step == max_epochs * train_batches
     assert trainer.fit_loop.epoch_loop._batches_that_stepped == max_epochs * train_batches
 
 
+def test_logging_step_loaded_correctly_pre_1_6_5(tmpdir):
+    trainer = Trainer(max_steps=1, default_root_dir=tmpdir)
```
Comment (carmocca) [inline tests/tests_pytorch/models/test_restore.py:267]: Should disable validation too

File: tests/tests_pytorch/models/test_restore.py [hunk 2]
```diff
@@ -255,13 +255,35 @@ class TestModel(BoringModel):
...
+    ckpt = torch.load(ckpt_path)
+    del ckpt["loops"]["fit_loop"]["epoch_loop.state_dict"]["_batches_that_stepped"]
+    torch.save(ckpt, ckpt_path)
+
+    class TestModel(BoringModel):
+        def on_train_start(self) -> None:
+            assert self.trainer.global_step == 1
+            assert self.trainer.fit_loop.epoch_loop._batches_that_stepped == 1
+
+    trainer = Trainer(max_steps=2, default_root_dir=tmpdir)
```
Comment (carmocca) [inline tests/tests_pytorch/models/test_restore.py:280]: Same here

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Nice catch !

File: tests/tests_pytorch/models/test_restore.py [hunk 3]
```diff
@@ -255,13 +255,35 @@ class TestModel(BoringModel):
...
     assert trainer.fit_loop.epoch_loop._batches_that_stepped == max_epochs * train_batches
 
 
+def test_logging_step_loaded_correctly_pre_1_6_5(tmpdir):
+    trainer = Trainer(max_steps=1, limit_val_batches=0, default_root_dir=tmpdir)
+    model = BoringModel()
+    trainer.fit(model)
+    ckpt_path = trainer.checkpoint_callback.best_model_path
+    ckpt = torch.load(ckpt_path)
+    del ckpt["loops"]["fit_loop"]["epoch_loop.state_dict"]["_batches_that_stepped"]
```
Comment (akihironitta) [inline tests/tests_pytorch/models/test_restore.py:272]:
```suggestion
    # the key "_batches_that_stepped" doesn't exist in checkpoints generated with <1.6.5
    del ckpt["loops"]["fit_loop"]["epoch_loop.state_dict"]["_batches_that_stepped"]
```

**Review by akihironitta — APPROVED** [review]
LGTM!

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 4]
```diff
@@ -293,7 +293,8 @@ def on_save_checkpoint(self) -> Dict:
     def on_load_checkpoint(self, state_dict: Dict) -> None:
         # cache the dataloader state dict until the dataloader objects are available
         self._dataloader_state_dict = state_dict.get("dataloader_state_dict")
-        self._batches_that_stepped = state_dict.get("_batches_that_stepped", 0)
+        # restore global step instead to make sure logging works correctly if checkpoints <v1.6.5 used to resume
+        self._batches_that_stepped = state_dict.get("_batches_that_stepped", self.global_step)
```
Comment (awaelchli) [inline src/pytorch_lightning/loops/epoch/training_epoch_loop.py:297]: This could be done without changing the logic if we had checkpoint migration #9396 :)

File: tests/tests_pytorch/models/test_restore.py [hunk 5]
```diff
@@ -255,13 +255,36 @@ class TestModel(BoringModel):
...
             assert self.trainer.global_step == first_max_epochs * train_batches
+            assert self.trainer.fit_loop.epoch_loop._batches_that_stepped == first_max_epochs * train_batches
 
     trainer.fit(TestModel(), ckpt_path=ckpt_path)
     assert trainer.current_epoch == max_epochs
     assert trainer.global_step == max_epochs * train_batches
     assert trainer.fit_loop.epoch_loop._batches_that_stepped == max_epochs * train_batches
 
 
+def test_logging_step_loaded_correctly_pre_1_6_5(tmpdir):
```
Comment (awaelchli) [inline tests/tests_pytorch/models/test_restore.py:266]: Have you considered writing a unit test instead? If so, why is it not possible?

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 4 above)
Comment (rohitgr7) [reply src/pytorch_lightning/loops/epoch/training_epoch_loop.py:297]: yep.. looks really relevant.

File: tests/tests_pytorch/models/test_restore.py (hunk 5 above)
Comment (rohitgr7) [reply tests/tests_pytorch/models/test_restore.py:266]: because this is how we restore global_step
https://github.com/Lightning-AI/lightning/blob/208c85013528704f4cc9727d44c95ebf5cde0b34/src/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L270

and by this point, the optimizer progress is not loaded, so it's not possible to assign the global step if we go with unit test approach.

Comment (rohitgr7) [reply tests/tests_pytorch/models/test_restore.py:266]: oh wait! just realized it might not work with manual optimization.

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 6]
```diff
@@ -264,10 +264,18 @@ def restore_loops(self) -> None:
...
         # it will be overwritten by the loop's state if it was also saved
-        optimizer_loop = fit_loop.epoch_loop.batch_loop.optimizer_loop
-        optimizer_loop.optim_progress.optimizer.step.total.completed = self._loaded_checkpoint["global_step"]
+        batch_loop = fit_loop.epoch_loop.batch_loop
+        if pl_module is None or pl_module.automatic_optimization:
+            batch_loop.optimizer_loop.optim_progress.optimizer.step.total.completed = self._loaded_checkpoint[
+                "global_step"
+            ]
+        else:
+            batch_loop.manual_loop.optim_step_progress.total.completed = self._loaded_checkpoint["global_step"]
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:277]:
```suggestion
        assert pl_module is not None
        loop = batch_loop.optimizer_loop if pl_module.automatic_optimization else batch_loop.manual_loop
        loop.optim_step_progress.total.completed = self._loaded_checkpoint["global_step"]
```

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 7]
```diff
@@ -264,10 +264,18 @@ def restore_loops(self) -> None:
...
 
         fit_loop = self.trainer.fit_loop
+        pl_module = self.trainer.lightning_module
+
         # set the `global_step` value for checkpoints before v1.6 without the progress tracking state.
         # it will be overwritten by the loop's state if it was also saved
-        optimizer_loop = fit_loop.epoch_loop.batch_loop.optimizer_loop
-        optimizer_loop.optim_progress.optimizer.step.total.completed = self._loaded_checkpoint["global_step"]
+        batch_loop = fit_loop.epoch_loop.batch_loop
+        if pl_module is None or pl_module.automatic_optimization:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:272]: Then let's make it an assertion :)

[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py]: done!
